### ISSUE TITLE
This tool is currently dangerous and should not be used until logic is added to preserve property order

### ISSUE THREAD

**Issue by ghost** [antonagestam/Squash#3]
This doesn't have any logic in it for keeping order in tact, which is _necessary_ when dealing with conflicting styles, which makes it worse than useless.

There's a good reason that nobody has already created something that does this -- it's a fairly complex problem to solve, much harder than simply throwing everything into a hash and reading it back out, which this seems to do.

For instance:

``` css
.foo { display: none;  }
.bar { display: block; }
.baz { display: none;  }
```

``` html
<div class="foo bar baz"></div>
```

the above div should be hidden, but if you're throwing the css, by rule, into a hash, it'll write out as:

``` css
.foo, .baz { display: none;  }
.bar       { display: block; }
```

thus breaking what you intended.

Basically, in order for this to "work", you need to write out the rule as soon as you find one that conflicts.  So in the above example, you _can not_ combine `foo` and `baz`'s `display` because there is a conflicting one in-between.  Similarly, if you have a bunch of `color` rules, you need to write the existing ones out as soon as new one is found.  Or for `border`'s (and `margin`, `padding`, `background`, etc) things get more complicated because you would need to write out any "partial border" styles as well, like `border-top`, etc..

At this point, you're stuck writing out the styles almost as often as they're defined, unless you happen to often have the same `property: value` pairs in succession.  You'll probably end up significantly increasing the size of the overall CSS because of this.

It _might_ be worthwhile to use this, and to check the size of the each "new" css rule against the size of writing out the "old" ones.  If the length of the selectors is less than the length of writing out the `property: value` rule multiple times, then it's good to go.  But my guess is that very few styles would ever meet this condition, things like `display: none`, `position: relative`, `overflow: hidden`, etc.  Things that are _very_ common.

The basic logic isn't even that complicated:

``` js
if( selectors.length + propVal.length + 2 /* { and } */ < propVal.length * numOccurrences ){
    // write out "selectors{ prop: val }"
    // remove "prop: val" from existing selectors
}
```

But it gets more significantly complicated when you start to account for selectors that would then become empty (thus changing the threshold for minimizing them like this), or when you want to combine multiple `prop: val`'s into one selector, e.g. if your optimizations want to write out something like:

``` css
.foo, .bar, .baz {
    border        : 5px;
    border-radius : 5px;
    color         : red;
}
```

Combining multiple properties into new selector groups is especially tricky since you don't, during parsing, know the optimal time to write out each property.  I imagine you'd have to "guess" by tracking how many characters it would use to add the selector now vs later, then go back and find the optimal point for each one.  Actually doing that is fairly complicated :)

And further, if this is a CSS min tool, it should be stripping white-space and line-breaks.  Otherwise people have to run generated code through _another_ minifier.

I've obviously thought about this a lot before now :)

**tl;dr** -- This is unusable as is (dangerous, even).  The idea has promise but is fairly tricky to implement properly, which is probably why nobody has yet.  Also, even if implemented properly the space savings would probably not be significant.

**Comment by corysimmons** [comment]
Agree with almost all your points but you could've been much nicer/less destructive during that excessive rant. =[

**Comment by ghost** [comment]
I don't think I was mean about it.  Direct, sure :)  And it's not a rant -- Most of that was directed at how to _fix_ this.  I'd actually like to see a tool like this and I tried to outline how it should be done properly.  But I think it's perfectly valid to warn that as it is right now, this tool should not be used.

**Comment by corysimmons** [comment]
No problem. Thanks for blocking out these points.

**Comment by antonagestam** [comment]
@zyklus I think you were nice enough and I'm aware of this fairly obvious problem. For this to work properly as is, you'd have to be very aware of how you write and use your selectors, maybe I should have pointed out the destructiveness of the tool in the README, but there's a PR for that now :)

Thanks for so clearly describing the problem and contributing with very valuable examples!
